app/controllers/user_sessions_controller.rb

```diff
@@ -6,75 +6,130 @@ def new
-          params[:user_session].delete(:username)
-          params[:openid] = username # pack up username for second runthrough
+      if signed_in?
+        if @identity.nil?
+          # If no identity was found, create a brand new one here
+          @identity = UserTag.create_with_omniauth(auth, current_user.id)
+          # The identity is not associated with the current_user so lets
+          # associate the identity
+          @identity.user = current_user
+          @identity.save
+          redirect_to root_url, notice: "Successfully linked to your account!"
+        elsif @identity.user == current_user
+          # User is signed in so they are trying to link an identity with their
+          # account. But we found the identity and the user associated with it
+          # is the current user. So the identity is already associated with
+          # this user. So let's display an error message.
+          redirect_to root_url, notice: "Already linked to your account!"
+        else
+          # User is signed in so they are trying to link an identity with their
+          # account. But we found the identity and a different user associated with it
+          # ,which is not the current user. So the identity is already associated with
+          # that user. So let's display an error message.
+          redirect_to root_url, notice: "Already linked to another account!"
         end
-        @user_session = UserSession.new(username: params[:user_session][:username], 
-                                        password: params[:user_session][:password], 
-                                        remember_me: params[:user_session][:remember_me])
-        saved = @user_session.save do |result|
-          if result
-            # replace this with temporarily saving pwd in session,
-            # and automatically saving it in the user record after login is completed
-            if current_user.crypted_password.nil? # the user has not created a pwd in the new site
-              flash[:warning] = I18n.t('user_sessions_controller.create_password_for_new_site')
-              redirect_to '/profile/edit'
-            else
-              flash[:notice] = I18n.t('user_sessions_controller.logged_in')
-              if session[:openid_return_to] # for openid login, redirects back to openid auth process
-                return_to = session[:openid_return_to]
-                session[:openid_return_to] = nil
-                redirect_to return_to
-              elsif session[:return_to]
-                return_to = session[:return_to]
-                session[:return_to] = nil
-                redirect_to return_to
-              elsif params[:return_to]
-                redirect_to params[:return_to]
+      else # not signed in
+        if @identity&.user.present?
+          # The identity we found had a user associated with it so let's
+          # just log them in here
+          self.current_user = @identity.user
```

Comment:
@jywarren here is the code. We have fetched the user which has same uid as provider by the google. Now we need to set the current user as this fetched user and set current session as this users sessions